python_modules/libraries/dagster-airlift/dagster_airlift/core/monitoring_job/event_stream.py

```diff
@@ -0,0 +1,295 @@
+from dagster._core.utils import make_new_run_id
+from dagster._record import record
+from dagster._time import datetime_from_timestamp
+from dagster_airlift.constants import DAG_ID_TAG_KEY, DAG_RUN_ID_TAG_KEY
+from dagster_airlift.core.airflow_defs_data import AirflowDefinitionsData
+from dagster_airlift.core.airflow_instance import AirflowInstance
+from dagster_airlift.core.monitoring_job.utils import (
+    get_dagster_run_for_airflow_repr,
+    structured_log,
+)
+from dagster_airlift.core.runtime_representations import DagRun, TaskInstance
+
+
+@record
+class DagRunStarted:
+    dag_run: DagRun
+
+    @property
+    def timestamp(self) -> float:
+        return self.dag_run.start_date.timestamp()
+
+    def persist_state(
+        self,
+        context: OpExecutionContext,
+        airflow_data: AirflowDefinitionsData,
+        airflow_instance: AirflowInstance,
+    ) -> None:
+        # If the run is already in Dagster, we don't need to do anything.
+        relevant_job_def = airflow_data.airflow_mapped_jobs_by_dag_handle.get(
+            self.dag_run.dag_handle
+        )
+        # Only if we have a relevant job def do we need to mess with creating runs.
+        if not relevant_job_def:
+            return
+        dagster_run_id = make_new_run_id()
+        context.instance.create_run_for_job(
+            run_id=dagster_run_id,
+            job_def=relevant_job_def,
+            tags={
+                DAG_RUN_ID_TAG_KEY: self.dag_run.run_id,
+                DAG_ID_TAG_KEY: self.dag_run.dag_id,
+                EXTERNAL_JOB_SOURCE_TAG_KEY: "airflow",
+            },
+            status=DagsterRunStatus.NOT_STARTED,
+            remote_job_origin=RemoteJobOrigin(
+                # We steal the repository origin from the original run.
+                # This allows the UI to link the run we're creating here back to the
+                # job in Dagster.
+                # It's not set in test contexts
+                repository_origin=context.run.remote_job_origin.repository_origin,
+                job_name=relevant_job_def.name,
+            )
+            if context.run.remote_job_origin
+            else None,
+        )
+
+        context.instance.report_dagster_event(
+            run_id=dagster_run_id,
+            dagster_event=DagsterEvent(
+                event_type_value="PIPELINE_START",
+                job_name=relevant_job_def.name,
+            ),
+        )
+
+        for asset_key in airflow_data.all_asset_keys_by_dag_handle[self.dag_run.dag_handle]:
+            context.instance.report_dagster_event(
+                run_id=dagster_run_id,
+                dagster_event=DagsterEvent.build_asset_materialization_planned_event(
+                    job_name=relevant_job_def.name,
+                    asset_materialization_planned_data=AssetMaterializationPlannedData(
+                        asset_key=asset_key,
+                    ),
+                    step_key="",
```

Comment:
guh -- probably not an issue but could we make some `__no_step__`-type step key for this?